composer/datasets/hparams.py

```diff
@@ -43,15 +47,119 @@ class DataloaderSpec(NamedTuple):
     split_fn: Callable[[Batch, int], List[Batch]] = _split_fn
 
 
-@dataclass
+@dataclasses.dataclass
 class DatasetHparams(hp.Hparams, ABC):
-    """Abstract base class for hyperparameters to initialize a dataset."""
+    """Abstract base class for hyperparameters to initialize a dataset.
+
+    Datasets MAY implement the following attributes. If a dataset does not implement an attribute,
+    a :class:`AttributeError` will be raised when trying access or set it.
+
+    Attributes:
+        synthetic (hp.Hparams, optional): If not None, then synthetic data will be returned. If None (the default),
+            then real data will be used.
+        num_total_batches (int, optional): If not None, the the dataloader from :meth:`initialize_object`
+            should have this length such that `len(dataloader) == num_total_batches`. Each epoch should yield
+            the same subset of samples.
+            
+            If this value is greater than the total number of samples in the dataset, then raises a :class:`ValueError`.
+
+            If this field is None, then the entire dataset will be iterated over.
+
+            This field is usally required if :attr:`synthetic` is not None.
+        shuffle (bool): Whether to shuffle the dataset.
+        drop_last (bool): Whether to drop the last batch.
+        datadir (str): The path to the data directory. Ignored if :attr:`synthetic` is not None.
+    """
+
+    def _common_field_getter(self, field_name: str) -> Any:
```

Comment:
I'm guessing this is something from yahp but what's the purpose of this and the `_common_field_setter`? 

i.e. why can't we just have dataclass fields like for other hparams?

Reply:
Yeah....so the logic behind this is that we want Exceptions to be raised when attempting to use common fields on datasets that do not implement common fields (for example, brats currently doesn't support synthetic data, so attempting to set synthetic on it should raise an exception.) If they were declared as fields on the parent, then it would not raise exceptions.

Reply:
Switched to using subclassing and isinstance checks -- much cleaner!